samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -262,3 +278,38 @@ def expect_minimum_argument_count!(arguments, min_count)
   false
 }
 
+#<ROUTINE name (arguments) #DECL body>
+ZIL_BUILTINS[:ROUTINE] = lambda { |arguments, context|
+  raise FunctionError, "ROUTINE must have a body!" unless arguments.length > 2
+  raise FunctionError, "ROUTINE must provide an argument list!" unless arguments.length > 1
+  raise FunctionError, "ROUTINE must provide an argument list! (" + arguments[1].class + ")" unless arguments[1].class == Syntax::List
+
+  name = arguments[0]
+  signature = arguments[1]
+  body = arguments[2..-1]
+
+  routine = ZIL::Routine.new(name, signature, body)
+  context.globals[name] = routine.method(:call)
+}
+
+ZIL_BUILTINS[:ASSIGNED?] = lambda { |arguments, context|
+  raise FunctionError, "ASSIGNED? requires at least one parameter!" unless arguments.length > 0
```

Comment:
I believe that with the changes I currently have, (not sure if I pushed them?) you don't need the routine name before the message, (also you can use my helpers 😉)

Reply:
Cleaned up `ASSIGNED?` `GASSIGNED?` with said helpers. 👍